src/hardware/pcspeaker.cpp

```diff
@@ -68,13 +96,49 @@ static bool SpeakerExists()
 	return spkr.chan != nullptr;
 }
 
+// Determines if the inbound wave is square by inspecting
+// current and previous states.
+static bool IsWaveSquare() {
+
+	// We have a square wave if .. 
+
+	// 1. PIT mode is enabled and PWM was or is turning on PIT mode.
+	if (spkr.pit_mode)
+		if (spkr.prev_mode == SPKR_PIT_ON || spkr.mode == SPKR_PIT_ON)
+			return true;
+
+	// 2. Speaker was in a neutral position and PWM is transitioning to PIT.
+	//    Transitioning being prev=OFF and current=ON, or prev=ON and current=OFF.
+	if (!spkr.prev_pos) {
+		const int pwm_sum = static_cast<int>(spkr.prev_mode) + static_cast<int>(spkr.mode);
```

Comment:
1. enum values being converted to `int` do not have associated values; perhaps they should have?
2. `SPKR_ON + SPKR_PIT_OFF` == `SPKR_OFF + SPKR_PIT_ON` - are these the two cases to be detected here?

Reply:
Hmm; yeah - this was all just me being lazy / clever using addition to catch the "xor" scenario.

Pushed another update, this time removing this enum-addition, and just spelling out each case.  Surely the compiler will optimize it into something smarter anyway :-), and we gain readability.

See what you think!

Reply:
@dreamer  - also wanted to thank you again for the careful feedback; the code is significantly improved!